### PR TITLE
[Datasets] Explicitly define Dataset-like APIs in DatasetPipeline class

### LINKED ISSUE #20888: [Bug] Cannot access member "map_batches" for type "DatasetPipeline[int]"
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Others

### What happened + What you expected to happen

  vscode and pyright report:

```
  12:13 - error: Cannot access member "map_batches" for type "DatasetPipeline[int]"
    Member "map_batches" is unknown (reportGeneralTypeIssues)
```

Because DatasetPipeline is a dynamically created method without type annotations.

This prevents vscode from:
1. being able to go to the definition of the method
1. being able to determine the return type, so autocomplete fails.

This adds friction during development. 

### Versions / Dependencies

ray 1.9.0

### Reproduction script

```
import ray

# Create a dataset and then create a pipeline from it.
base = ray.data.range(1000000)
print(base)
# -> Dataset(num_blocks=200, num_rows=1000000, schema=<class 'int'>)
pipe = base.window(blocks_per_window=10)
print(pipe)
# -> DatasetPipeline(num_windows=20, num_stages=1)

# Applying transforms to pipelines adds more pipeline stages.
pipe = pipe.map_batches(lambda batch: [v * 2 for v in batch])
print(pipe)
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

Rework of https://github.com/ray-project/ray/pull/26394, which gets reverted due to unit test failure.

## Related issue number

Closes https://github.com/ray-project/ray/issues/20888
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_dataset_pipeline.py [hunk 1]
```diff
@@ -449,6 +491,25 @@ def test_count_sum_on_infinite_pipeline(ray_start_regular_shared):
...
+    pipe = ray.data.range(12).window(blocks_per_window=3).sort_each_window()
+    assert pipe.take() == list(range(12))
+
+    pipe = (
+        ray.data.range(12)
+        .window(blocks_per_window=12)
+        .sort_each_window(descending=True)
+    )
+    assert pipe.take() == [11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
+
```
Comment (c21) [inline python/ray/data/tests/test_dataset_pipeline.py:504]: This is the only difference compared to https://github.com/ray-project/ray/pull/26394, rest of change in this PR are the same. To make unit test stable, we'd better only generate 1 window and then sort.

**Review by scv119 — APPROVED** [review]

File: python/ray/data/tests/test_dataset_pipeline.py (hunk 1 above)
Comment (jianoaix) [reply python/ray/data/tests/test_dataset_pipeline.py:504]: Hm, why having blocks_per_window=3 will fail? In this case, it's same as ray.data.range(12).repeat(1) and then sort (i.e. no real windowing).

Comment (c21) [reply python/ray/data/tests/test_dataset_pipeline.py]: Actuallly after debugging with @clarkzinzow, found out parallelism (and number of blocks) in unit test is changed after https://github.com/ray-project/ray/pull/25883 . Previously the number of blocks / parallelism here is 12, right now is 2, due to in pytest with ray_start_regular_shared cluster fixture, available CPUs is 1, so parallelism becomes 2* CPUs, i.e. 2 here (given data size is very small here). So we should explicitly set number of parallelism here.

File: python/ray/data/tests/test_dataset_pipeline.py [hunk 2]
```diff
@@ -556,6 +598,29 @@ def test_count_sum_on_infinite_pipeline(ray_start_regular_shared):
...
+def test_sort_each_window(ray_start_regular_shared):
+    pipe = (
+        ray.data.range(12, parallelism=12)
+        .window(blocks_per_window=3)
+        .sort_each_window()
+    )
+    assert pipe.take() == list(range(12))
+
+    pipe = (
+        ray.data.range(12, parallelism=12)
```
Comment (c21) [inline python/ray/data/tests/test_dataset_pipeline.py:610]: add `parallelism=12` here.

File: python/ray/data/tests/test_dataset_pipeline.py [hunk 3]
```diff
@@ -556,6 +598,29 @@ def test_count_sum_on_infinite_pipeline(ray_start_regular_shared):
...
+        .window(blocks_per_window=3)
+        .sort_each_window(descending=True)
+    )
+    assert pipe.take() == [2, 1, 0, 5, 4, 3, 8, 7, 6, 11, 10, 9]
+
+    pipe = (
+        ray.data.range(12, parallelism=12)
```
Comment (c21) [inline python/ray/data/tests/test_dataset_pipeline.py:617]: add `parallelism=12` here.

**Review by clarkzinzow — APPROVED** [review]
LGTM!

File: python/ray/data/tests/test_dataset_pipeline.py (hunk 1 above)
Comment (jianoaix) [reply python/ray/data/tests/test_dataset_pipeline.py]: Ah, I see

**Review by jianoaix — APPROVED** [review]

**Comment by c21** [comment]
Thank you @clarkzinzow, @jianoaix and @scv119 for review!

**Pull request merged**